sys/usb/usbus/msc/msc.c

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (C) 2019-2021 Mesotic SAS
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usbus_msc
+ * @{
+ *
+ * @author  Dylan Laduranty <dylan.laduranty@mesotic.com>
+ * @}
+ */
+
+#include "usb/descriptor.h"
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+
+#include "usb/msc.h"
+#include "usb/usbus/msc.h"
+#include "usb/usbus/msc/scsi.h"
+#include "board.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+#include "mtd.h"
+
+#ifndef MTD_MSC
+/* TODO: add support for multiple MTD devices */
+#define MTD_MSC MTD_0
+#endif
+
+/* Internal buffer to handle size difference between MTD layer and USB
+   endpoint size as some MTD implementation (like mtd_sdcard) doesn't allow
+   endpoint size transfer */
+static unsigned char _buff[USBUS_MSC_BUFFER_SIZE];
+static unsigned char _ep_out_buf[CONFIG_USBUS_EP0_SIZE];
+static unsigned char _ep_in_buf[CONFIG_USBUS_EP0_SIZE];
+
+/* Internal handler definitions */
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                           usbus_event_usb_t event);
+
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup);
+
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+
+static const usbus_handler_driver_t msc_driver = {
+    .init = _init,
+    .event_handler = _event_handler,
+    .control_handler = _control_handler,
+    .transfer_handler = _transfer_handler,
+};
+
+static const usbus_descr_gen_funcs_t _msc_descriptor = {
+    .fmt_post_descriptor = NULL,
+    .fmt_pre_descriptor = NULL,
+    .len = {
+        .fixed_len =  0,
+    },
+    .len_type = USBUS_DESCR_LEN_FIXED,
+};
+
+static void _write_xfer(usbus_msc_device_t *msc)
+{
+    /* Check if we have a block to read and transfer */
+    if (!msc->block_nb) {
+        return;
+    }
+
+    size_t len;
+    /* Retrieve incoming data */
+    usbdev_ep_get(msc->ep_out->ep, USBOPT_EP_AVAILABLE, &len, sizeof(size_t));
+
+    /* Update offset for page buffer */
+    msc->block_offset += len;
+    /* Decrement whole len */
+    msc->cmd.len -= len;
+
+    /* buffer is full, write it and point to new block if any */
+    if (msc->block_offset >= (MTD_MSC->page_size * msc->pages_per_vpage)) {
+        mtd_write_page(MTD_MSC, msc->buffer, msc->block * msc->pages_per_vpage,
```

Comment:
If you write whole pages you might also use `mtd_write_page_raw`.
Then you could even use the per-mtd device sector sized work buffer instead of allocating your own

Reply:
The problem with this approach is for devices with more than 4096 bytes as page size. I'll be force to report 4096 as page size so Host can be happy but I'll need to write less than a page.